src/utils/validateFunction.ts

```diff
@@ -0,0 +1,102 @@
+import fs from 'fs-extra';
+import path from 'path';
+import fetch from 'node-fetch';
+import { Validator, ValidatorResult, ValidationError } from 'jsonschema';
+
+type FunctionDefinition = {
+  name?: string;
+  [other: string]: unknown;
+};
+
+export type Schema = {
+  $id: string;
+};
+
+export type Config = {
+  schemaUrl: string;
+  functionSchemaPath: string;
+};
+
+const fetchFunction = async (
+  functionPath: string,
+): Promise<FunctionDefinition> => {
+  const filePath = path.join(functionPath, 'function.json');
+  try {
+    const exists = await fs.pathExists(filePath);
+    if (!exists) throw new Error(`file not found ${filePath}`);
+    return (await fs.readJSON(filePath)) as FunctionDefinition;
+  } catch (err) {
+    throw new Error(`could not load json from ${functionPath}: ${err}`);
+  }
+};
+
+const fetchRemoteSchema = async (schemaUrl: string): Promise<Schema> => {
+  const res = await fetch(schemaUrl);
+  const json = await res.json();
+  return json as Schema;
+};
+
+const importNextSchema = async (
+  validator: Validator,
+  schemaId: string,
+): Promise<Validator> => {
+  const schemaJSON = await fetchRemoteSchema(schemaId);
+  validator.addSchema(schemaJSON, schemaId);
+
+  const nextSchemaId = validator.unresolvedRefs.shift();
+  if (!nextSchemaId) {
+    return validator;
+  }
+  return importNextSchema(validator, nextSchemaId);
+};
+
+const importSchema = async (
+  validator: Validator,
+  config: Config,
+): Promise<Validator> => {
+  const functionSchemaUrl = config.schemaUrl + config.functionSchemaPath;
+  return importNextSchema(validator, functionSchemaUrl);
+};
+
+const functionValidator = async (config: Config): Promise<Validator> => {
+  const validator = new Validator();
+
+  return importSchema(validator, config);
+};
+
+const validateFunctionDefinition = async (
```

Comment:
There is no async code here, so I think we can remove the `async`

Reply:
this can also be a custom type guard to do the "conversion". All we need to do is make the return type [a type predicate](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates).

```ts
const validateFunctionDefinition = (validator, parsedJson: object): parsedJson is FunctionDefinition => {

    ///....
} 
```

if the function above returns true, typescript will follow the instruction at the end

so now you can just

```ts

const parsedJson = fetchFunction(pathToFunction);
if (!validateFunctionDefinition(parsedJson)) throw new Error('your input is invalid');
 
 // from here on typescript no longer thinks `parsedJson` is just
 // object, it now knows it is a `FunctionDefinition` Object.

const functionDefinition: FunctionDefinition = parsedJson // this is now allowed

```